frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -104,297 +178,465 @@ const defaultValueFor = (field: OperandField) => {
+            fieldListsAccumulator[index] = [...(fieldListsAccumulator[index] || []), field];
+          }
+          return fieldListsAccumulator;
+        },
+        [],
+      ),
+    }));
+  }, [arrayFields]);
+
+  // Create memoized fieldGroups. Map to array of fields grouped by fieldGroup name so that sorting
+  // is easy.
+  const fieldGroups = React.useMemo(() => {
+    const groupedByName = _.groupBy(groupFields, (field) => {
+      const { groupName } = parseGroupDescriptor(field);
+      return groupName;
+    });
 
-  const [formValues, setFormValues] = React.useState<FormValues>({
-    'metadata.name': _.get(sample, 'metadata.name', 'example'),
-    'metadata.labels': SelectorInput.arrayify(_.get(sample, 'metadata.labels')) || [],
-    ...defaultFormValues,
-    ...sampleFormValues,
-  });
+    return _.map(groupedByName, (fieldList, groupName) => ({
+      groupName,
+      fieldList,
+    }));
+  }, [groupFields]);
 
-  const [error, setError] = React.useState<string>();
-  const [formErrors, setFormErrors] = React.useState<FieldErrors>({});
-  const [arrayFieldGroups, setArrayFieldGroupList] = React.useState<
-    Set<SpecCapability.arrayFieldGroup>
-  >(getArrayFieldGroups());
-
-  const [k8sObj, setK8sObj] = React.useState<K8sResourceKind>(sample);
-  React.useEffect(() => {
-    setK8sObj((current) => {
-      const specValues = fields.reduce((usedFormValues, field) => {
-        const formValue = _.get(usedFormValues, field.path);
-        if (_.isEqual(formValue, defaultValueFor(field)) || _.isNil(formValue)) {
-          return _.omit(usedFormValues, field.path);
-        }
-        return usedFormValues;
-      }, _.omit(formValues, ['metadata.name', 'metadata.labels']));
+  const updateFormValues = React.useCallback((path: string, value: any) => {
+    setFormValues((current) => {
       const next = {
-        apiVersion: apiVersionForModel(operandModel),
-        kind: operandModel.kind,
-        metadata: {
-          namespace,
-          name: formValues['metadata.name'],
-          labels: SelectorInput.objectify(
-            formValues['metadata.labels'],
-          ) as ObjectMetadata['labels'],
-          annotations: _.get(sample, 'metadata.annotations', {}),
-        },
-        spec: _.reduce(
-          specValues,
-          (spec, value, path) => _.set(spec, path, value),
-          _.get(sample, 'spec', {}),
-        ),
+        ...current,
+        [path]: value,
       };
       if (_.isEqual(current, next)) {
         return current;
       }
-      onChange(next);
       return next;
     });
-  }, [fields, formValues, namespace, onChange, operandModel, sample]);
+  }, []);
 
-  const updateFormValues = (path: _.PropertyPath, value: any) => {
-    setFormValues((current) => {
-      const next = _.set(_.cloneDeep(current), path, value);
-      return _.isEqual(current, next) ? current : next;
-    });
-  };
+  // Build K8s object from current state
+  const formData = React.useMemo(() => {
```

Comment:
Renamed the k8sObj state to formData and changed to useMemo, since the previous implementation was essentially doing the same thing using useState and useEffect.

Reply:
I also changed the logic a little bit in this function. Before, if the 'buffer' (previously named 'sample') prop had a spec property with an array value, and the length of that array was greater than the corresponding array from the form, they were getting merged together and the array elements from the form would overwrite the buffer values in the same index, but the elements at higher indexes would still be present. So you would essentially get duplicate array elements in the form. Now, the new array from the form just replaces the one in the buffer.